boards/shared/Src/App/App_SharedSignal.c

```diff
@@ -1,135 +1,67 @@
 
-    return signal;
-}
-
-void App_SharedSignal_Destroy(struct Signal *signal)
-{
-    free(signal);
-}
+    App_Timer_InitTimer(signal->entry_timer, entry_time);
+    App_Timer_InitTimer(signal->exit_timer, exit_time);
 
-uint32_t App_SharedSignal_GetEntryLastTimeLowMs(const struct Signal *signal)
-{
-    return signal->entry_last_time_low_ms;
-}
-
-uint32_t App_SharedSignal_GetEntryLastTimeHighMs(const struct Signal *signal)
-{
-    return signal->entry_last_time_high_ms;
-}
-
-uint32_t App_SharedSignal_GetExitLastTimeLowMs(const struct Signal *signal)
-{
-    return signal->exit_last_time_low_ms;
-}
-
-uint32_t App_SharedSignal_GetExitLastTimeHighMs(const struct Signal *signal)
-{
-    return signal->exit_last_time_high_ms;
-}
-
-bool App_SharedSignal_IsCallbackTriggered(const struct Signal *const signal)
-{
-    return signal->is_callback_triggered;
+    return signal;
 }
 
-void App_SharedSignal_Update(struct Signal *signal, uint32_t current_time_ms)
+SignalState App_SharedSignal_Update(struct Signal *signal, bool entry_condition_high, bool exit_condition_high)
 {
-    if (signal->is_entry_condition_high(signal->world))
-    {
-        signal->entry_last_time_high_ms = current_time_ms;
-    }
-    else
-    {
-        signal->entry_last_time_low_ms = current_time_ms;
-    }
-
-    if (signal->is_exit_condition_high(signal->world))
-    {
-        signal->exit_last_time_high_ms = current_time_ms;
-    }
-    else
-    {
-        signal->exit_last_time_low_ms = current_time_ms;
-    }
+    TimerState entry_timer_state = App_Timer_RunIfCondition(signal->entry_timer, entry_condition_high);
+    TimerState exit_timer_state  = App_Timer_RunIfCondition(signal->exit_timer, exit_condition_high);
 
     if (!signal->is_callback_triggered)
     {
-        const uint32_t time_entry_condition_high_ms = current_time_ms - signal->entry_last_time_low_ms;
-
-        if (time_entry_condition_high_ms >= signal->callback.entry_condition_high_duration_ms)
+        if (entry_timer_state == TIMER_STATE_EXPIRED)
         {
-            signal->callback.function(signal->world);
+            signal->state                 = SIGNAL_ENTRY_HIGH;
             signal->is_callback_triggered = true;
         }
     }
     else
     {
-        const uint32_t time_exit_condition_high_ms = current_time_ms - signal->exit_last_time_low_ms;
-
-        if (time_exit_condition_high_ms >= signal->callback.exit_condition_high_duration_ms)
-        {
+        // keep running callback function for (exit_condition_high_duration_ms) time.
+        if (exit_timer_state == TIMER_STATE_EXPIRED)
             signal->is_callback_triggered = false;
-        }
         else
-        {
-            signal->callback.function(signal->world);
-        }
+            signal->state = SIGNAL_EXIT_HIGH;
     }
+    return signal->state;
+}
+
+// Getters
+bool App_SharedSignal_IsCallbackTriggered(const struct Signal *const signal)
+{
+    return signal->is_callback_triggered;
+}
+
+// Destructor
```

Comment:
nit: I don't like comments like these, they don't give you much useful info (and there's already a better comment in the .h file for destroy)
nit 2: We usually order the functions "..._Create", "..._Destroy", then the rest

Reply:
Fixed with bf889d9c8cd8a129c8f54c7c1bcc674ce2fb78c3